File: gommo-frontend/src/modules/payroll/lib/payroll-run.mapper.ts
import type { PayrollRun, PayrollRunCreateDto } from "@/modules/payroll/dto/payroll-run.dto";
import { currentMonthReferenceDate, isoToMonthBr, normalizeMonthIso } from "@/shared/lib/input/date";

function toDateField(value?: string): string | undefined {
    if (!value) return undefined;
    return value.slice(0, 10);
}

export function formatPayrollReference(referenceDate: string): string {
    return isoToMonthBr(referenceDate) || "—";
}

export function payrollrunToFormDto(entity: PayrollRun): PayrollRunCreateDto {
    return {
        referenceDate: normalizeMonthIso(entity.referenceDate ?? ""),
        payrollStatus: entity.payrollStatus === "DRAFT" ? "OPEN" : entity.payrollStatus,
        openedAt: toDateField(entity.openedAt),
        closedAt: toDateField(entity.closedAt),
        processedAt: entity.processedAt,
        notes: entity.notes,
    };
}

export const emptyPayrollRunForm = (): PayrollRunCreateDto => ({
    referenceDate: currentMonthReferenceDate(),
    payrollStatus: "OPEN",
});

File: gommo-backend/src/main/java/br/com/gommo/modules/payroll/mapper/PayrollRunMapper.java
package br.com.gommo.modules.payroll.mapper;

import br.com.gommo.modules.payroll.dto.PayrollRunRequestDto;
import br.com.gommo.modules.payroll.dto.PayrollRunResponseDto;
import br.com.gommo.modules.payroll.entity.PayrollRun;
import br.com.gommo.modules.payroll.entity.PayrollStatusEnum;
import java.time.LocalDate;
import org.springframework.stereotype.Component;

@Component
public class PayrollRunMapper {

    public PayrollRun toEntity(PayrollRunRequestDto dto) {
        return PayrollRun.builder()
                .referenceDate(normalizeReferenceDate(dto.getReferenceDate()))
                .payrollStatus(dto.getPayrollStatus() != null ? dto.getPayrollStatus() : PayrollStatusEnum.OPEN)
                .openedAt(dto.getOpenedAt())
                .closedAt(dto.getClosedAt())
                .processedAt(dto.getProcessedAt())
                .notes(dto.getNotes())
                .build();
    }

    public void updateEditableFields(PayrollRun entity, PayrollRunRequestDto dto) {
        entity.setReferenceDate(normalizeReferenceDate(dto.getReferenceDate()));
        entity.setNotes(dto.getNotes());
    }

    public void updateEntity(PayrollRun entity, PayrollRunRequestDto dto) {
        updateEditableFields(entity, dto);
    }

    public PayrollRunResponseDto toResponse(PayrollRun entity) {
        return PayrollRunResponseDto.builder()
                .id(entity.getId())
                .code(entity.getCode())
                .status(entity.getStatus())
                .referenceDate(entity.getReferenceDate())
                .payrollStatus(entity.getPayrollStatus())
                .openedAt(entity.getOpenedAt())
                .closedAt(entity.getClosedAt())
                .processedAt(entity.getProcessedAt())
                .notes(entity.getNotes())
                .createdAt(entity.getCreatedAt())
                .updatedAt(entity.getUpdatedAt())
                .build();
    }

    static LocalDate normalizeReferenceDate(LocalDate referenceDate) {
        if (referenceDate == null) {
            return null;
        }
        return referenceDate.withDayOfMonth(1);
    }
}
